<a href="https://colab.research.google.com/github/jcdumlao14/My-Notes-ml-zoomcamp2022/blob/main/week_1_2_notes_Intro_to_NumPy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Machine Learning Zoomcamp**

##**1.7 Introduction to NumPy**

### Plan:

* Creating arrays
* Multi-dementional arrays
* Randomly generated arrays
* Element-wise operations
  * Comparison operations
  * Logical operations

* Summarizing operations

In [ ]:
import numpy as np

* import the package: import num import and name of the package import NumPy and we use an alias as np

In [ ]:
np

<module 'numpy' from '/usr/local/lib/python3.7/dist-packages/numpy/__init__.py'>

* NumPy we can just write np

# **Creating arrays**

In [ ]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

* Start creating arrays

* Simplest method for that is called zeros simplest function so it takes one argument this is the size  of the array and it just creates an array of this size of the specified size with zero(10)

In [ ]:
np.ones(10)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

* Similar that we called once and create size(10)

In [ ]:
np.full(10, 2.5)

array([2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5])

* The you use a function called **full** an array with some arbitrary number, then first specify the size of the array (10, 2.5), you specify the element you want to fill the array with in case 2.5 and array of size 10, these how you create arrays with NumPy


In [ ]:
a = np.array([1, 2, 3, 5, 7, 12])
a

array([ 1,  2,  3,  5,  7, 12])

* If you want to convert a python list so **1,2,3,5,7,12** so, for example, we have this python list and create an array from this use **a** function called **array** and then pass the list the argument and then the result is an array, now we called the **array a**


In [ ]:
a[2]

3

* If we want to access an element of this array the element three with, here we just write index two so remember indexing in python starts with zero, and then this is how we access the third element, like in this case element number three

* Likewise if we want to change this to something else, let's say we want to change it to replace three with 10, so just use the assignment operator

In [ ]:
a[2] = 10

In [ ]:
a

array([ 1,  2, 10,  5,  7, 12])

* And hereafter executing code and here we have 10 instead of three, and the useful function for creating is on the range, create a range from 0 to 9, so 10 is like the last element not in its exclusive, so basically up to and by default its fro 0 to 10 exclusive, so in this case, it's from 0 to 9 and let's say if we start another number, not with three, but just put another parameter it creates range from three to nine this is very similar a python function called range, so it's doing the same thing except instead of giving us an iterator from python it gives as a NumPy array.

In [ ]:
np.arange(3, 10)

array([3, 4, 5, 6, 7, 8, 9])

In [ ]:
np.linspace(0, 100, 11)

array([  0.,  10.,  20.,  30.,  40.,  50.,  60.,  70.,  80.,  90., 100.])

* Another useful function in NumPy is called **linspace** which creates an array filled with a number between the first parameter and the second parameter, so let's say we want to create an array with numbers from 0 to 1, so the first element of this array is 0 the last is 1 and then create an array of size 10 of or size 11 like specify here and if fills everything between these two numbers and of course like instead of one we can put anything you want here, we created a one-dimensional array and multi-dimensional arrays.

# **Multi-dementional arrays**

In [ ]:
np.zeros((5, 2))

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

* So let's say if we want to create a **multi-dimensional array** we use the same function previously zeros, but here we can specify the dimensions of the array, so we can say that we want to create an array with five rows and two columns, so the first number here is rows takes second is columns and then it creates an array with five columns, **"5 rows and 2 columns"**. two raws filled with zeros.

In [ ]:
a = np.array([1,2,3])
a

array([1, 2, 3])

* The same way we can create an array from python list, so let's say we have an array, we have **a** list of list in python, one two three the four set the seven eight nine, so this is a list of python we pass it to array function it creates a two-dimensional array for us let's called **n** and again if we want to access an element of this array

In [ ]:
n = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

In [ ]:
n

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

* If you want to acces elements of this array this is basically the same as one dimensional except, we have now, indices and it's indexed wiht the it starts with zero, so zero one two for rows and zero one two for columns and if we want to now let's say access.

In [ ]:
n[0,1]

2

* This element number is from row number zero and column number one, so we use the first we specify zero and then one so it gives us this element

In [ ]:
n[0, 1] = 20

In [ ]:
n

array([[ 1, 20,  3],
       [ 4,  5,  6],
       [ 7,  8,  9]])

* First we specify zero and then one so, it gives us this element we can use the assignment operator, so if we want to replace this 2 with 20, so we just use the assignment operator and then we see that it changes, so now we change this element to 20.

In [ ]:
n[2] = [1, 1, 1]

In [ ]:
n

array([[1, 2, 3],
       [4, 5, 6],
       [1, 1, 1]])

* If we only want to access the rows then instead of passing two indices here we can pass one and let's say we want to get the zero row to say from number one row number two and let's say if we want to get the zero row, the row number one or row number two, so this how we access entire row and again we can use the assignment operator, let's say if we want to rewrite the entire row, we can say the last row should be just once we can say that we can assign using python list on NumPy array like the only condition that it should match the dimension here.

In [ ]:
n[:, 1] = [20, 5, 1]

In [ ]:
n

array([[ 1, 20,  3],
       [ 4,  5,  6],
       [ 1,  1,  1]])

* Some we can do with if we want to access a column we can get the entire column if we want and for that, we need to use only to specify only second argument but, let's say if we want to get the second array a second column we use one, but we cannot leave it empty, so we need to put something and we put the column operator meaning we want to get all the rows, but just keep the column indexed by the number we get this second column here right.

In [ ]:
n[:,2]

array([0, 1, 2])

* And this column operator means return all the rows so we can just use that as if returns everything and let's say if we want to set the last column this is what we get ([3,6,1]).

In [ ]:
n[:, 2] = [0, 1, 2]

In [ ]:
n

array([[ 1, 20,  0],
       [ 4,  5,  1],
       [ 1,  1,  2]])

* And likewise we can assign something else to this let's say if we want to reassign it to zero one or two we do that and then when we look again at our array we see that we managed to change it. these are two-dimensional arrays.

# **Randomly generated arrays**

In [ ]:
np.random.seed(2)
100 * np.random.rand(5, 2)

array([[43.59949021,  2.59262318],
       [54.96624779, 43.53223926],
       [42.03678021, 33.0334821 ],
       [20.4648634 , 61.92709664],
       [29.96546737, 26.68272751]])

* We create a Randomly generated array we can use this random package in NumPy, so let's generate a two-dimensional array of size five rows and two columns, we just specify the number of rows, and number of columns and it gives us an array with random numbers between zero and one so this is like a uniform distribution, standard uniform distribution, so every time we execute this cell we get a different number but if we want to make the results reproducible

In [ ]:
np.random.seed(10)
np.random.randn(5, 2)

array([[ 1.3315865 ,  0.71527897],
       [-1.54540029, -0.00838385],
       [ 0.62133597, -0.72008556],
       [ 0.26551159,  0.10854853],
       [ 0.00429143, -0.17460021]])

* If you run this code we get the same numbers it's possible we do this by fixing the random sets because the numbers that are not random are student pseudorandom so they are generated then, we can set the seed of this generator, and if we do this the sequence of random numbers is produced on my computer and will be the same

In [ ]:
np.random.seed(2)
np.random.randn(5, 2)

array([[-0.41675785, -0.05626683],
       [-2.1361961 ,  1.64027081],
       [-1.79343559, -0.84174737],
       [ 0.50288142, -1.24528809],
       [-1.05795222, -0.90900761]])

* These numbers are random but every time we execute this cell the results are the same right because we fixed the random sets like the seed number here for the random generation generator that's why if you execute you probably will have the same one, a bit varies depending on the version of NumPy or operating system you have but n general the sequence you get the matrix you get should be the same as I get.

* So here we sample from the uniform distribution, standard uniform distribution there are many different distributions, so let's say if we want to draw numbers from the standard normal distribution, we just add an **n** here so it's around, around **n** meaning and normal and then the distribution looks a bit different.

In [ ]:
np.random.seed(2)
np.random.randint(low=0, high=100, size=(5, 2))

array([[40, 15],
       [72, 22],
       [43, 82],
       [75,  7],
       [34, 49]])

* For integers like, let's say if we want to have numbers between 0 and 100 so one simple thing we can do is we can multiply every number here by 100 and actually, since this is numpy array we can just use that and this is something will show your in this section you can multiply every element by 100 and then you will set instead of having number between 0 and 1, you will have numbers 0 between 0 and 100 but if want there is a also a function that you can use for generating random integers, so you can specify here the low = lowest numbers and the highest number that is possible to set I think it's one is includes, included this one is probably excluded so its between 0 and  99 and then you can say that I want to have five rows and two columns and it will generate you a matrix and a two-dimensional array like that with random integers.

# **Element-wise operations**

In [ ]:
a = np.arange(5)
a

array([0, 1, 2, 3, 4])

* So here we did this multiplication let me explain how it works.
* Now we just create an array first a range of five so it's number from between zero and five, four in this case

In [ ]:
a + 1

array([1, 2, 3, 4, 5])

* We can do let's say what if we want to add **a**  number to this array so we want to add one to all the elements of this array was a usual, python list then we won't be able to do this we would meet to write a loop and then add number element by numpy makes it easier for us we can just write plus one and it adds an element to each number here.

In [ ]:
a * 2

array([0, 2, 4, 6, 8])

In [ ]:
b = (10 + (a * 2)) ** 2 / 100


In [ ]:
b

array([1.  , 1.44, 1.96, 2.56, 3.24])

* If we want to multiply by two then every element of this array you can see so 0 is multiplied by 2, so basically every elements is multiplied by this number 2. what we just multiplied everything by 100 and of course we can divide it subtact do everything we wan and not only that we can also chain them so let's say we can first divide, multiply by 2 and then add 10 and even let's  say take power of it we can change this equation anyway, like, whatever we want, let's save it to a different array so now we have array one array a and three b what we can do also and combine these two arrays so can do element wise sum, so we can take add element wise these two array for the first element we would compute the second for the second element we would compute the sum, let's say the last element here is 3.24, which is the sum of these two, or the second 2.56, 1.44 is the same of these two does the sum element twice.

In [ ]:
a / b + 10

array([10.        , 10.69444444, 11.02040816, 11.171875  , 11.2345679 ])

* Or divide, we can do anything we want here right, and then again we can chain it with all these different operations.

# **Comparison operations**

In [ ]:
a

array([0, 1, 2, 3, 4])

* let's print **a**
* Comparision Operations are also element-wise operations, so let's say we want to see if the element of this array which element of this array is higher or equal to two.

In [ ]:
a >= 2

array([False, False,  True,  True,  True])

* First one is less than to 0 is false, then second is less than the 1 is false and the two but this one is higher or equal to two because it's two and then the same is true for the rest.

In [ ]:
b

array([1.  , 1.44, 1.96, 2.56, 3.24])

In [ ]:
a > b

array([False, False,  True,  True,  True])

* We can do it other arrays, so let's say we can see which elements of array **a** are greater than the elements array **b**, like this one is actually smaller the second one is also smaller because zero is more than one, one is smaller than one point 44 but two is bigger then two is bigger and three is bigger , false, false, true, true, you know for compare elements of multiple arrays

In [ ]:
a[a > b]

array([2, 3, 4])

* We can see which elements of **a** are actually greater  **b** if we write something like this and execute it and so what happen here. these statement **a > b** returns a boolean array, so it returns an array with false, false,true,true, when we do these **a > b** it basically returns all the element of array **a** for which this condition is true.

# **Summarizing operations**

In [ ]:
a

array([0, 1, 2, 3, 4])

In [ ]:
a.std()

1.4142135623730951

In [ ]:
a.max()

4

In [ ]:
a.min()

0

In [ ]:
a.sum()

10

In [ ]:
n.min()

0

In [ ]:
n.std()

5.89622184952071

In [ ]:
n.sum()

35

* There ar different kinds of operations that we can call summarizing operations because instead of returning another array, return a single number so an example of this is **a min** operation **a** smallest number in this case it's zero, you can use **a.maximum** number.

* also cumpute the sum of all the arrays
* And you can compute the average number **a.mean()**
* And you can compute the standard deviation **a.std()**
* You also work for two dimensional, let's say n to compute the sum **n.sum()** of all the elements
* find the minimal number of **n.min()**

## **Next**

Linear algebra refresher
